src/Microsoft.AspNet.Mvc.Razor/Precompilation/CodeAnalysisSymbolBasedTypeInfo.cs

```diff
@@ -0,0 +1,251 @@
+    {
+        private static readonly System.Reflection.TypeInfo OpenGenericDictionaryTypeInfo =
+            typeof(IDictionary<,>).GetTypeInfo();
+        private readonly CodeAnalysisSymbolLookupCache _symbolLookup;
+        private readonly ITypeSymbol _type;
+        private string _fullName;
+        private List<IPropertyInfo> _properties;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CodeAnalysisSymbolBasedTypeInfo"/>.
+        /// </summary>
+        /// <param name="propertySymbol">The <see cref="IPropertySymbol"/>.</param>
+        /// <param name="symbolLookup">The <see cref="CodeAnalysisSymbolLookupCache"/>.</param>
+        public CodeAnalysisSymbolBasedTypeInfo(
+            [NotNull] ITypeSymbol type,
+            [NotNull] CodeAnalysisSymbolLookupCache symbolLookup)
+        {
+            _symbolLookup = symbolLookup;
+            _type = type;
+        }
+
+        /// <inheritdoc />
+        public string FullName
+        {
+            get
+            {
+                if (_fullName == null)
+                {
+                    _fullName = GetFullName(_type);
+                }
+
+                return _fullName;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool IsAbstract => _type.IsAbstract;
+
+        /// <inheritdoc />
+        public bool IsGenericType
+        {
+            get
+            {
+                return _type.Kind == SymbolKind.NamedType &&
+                    ((INamedTypeSymbol)_type).IsGenericType;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool IsNested => _type.ContainingType != null;
+
+        /// <inheritdoc />
+        public bool IsPublic => _type.DeclaredAccessibility == Accessibility.Public;
+
+        /// <inheritdoc />
+        public string Name => _type.MetadataName;
+
+        /// <inheritdoc />
+        public IEnumerable<IPropertyInfo> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = PopulateProperties();
+                }
+
+                return _properties;
+            }
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<TAttribute> GetCustomAttributes<TAttribute>()
+            where TAttribute : Attribute
+        {
+            return CodeAnalysisAttributeUtilities.GetCustomAttributes<TAttribute>(_type, _symbolLookup);
+        }
+
+        /// <inheritdoc />
+        public bool ImplementsInterface(System.Reflection.TypeInfo interfaceType)
+        {
+            var interfaceSymbol = _symbolLookup.GetSymbol(interfaceType);
+            return _type.AllInterfaces.Any(implementedInterface => implementedInterface == interfaceSymbol);
+        }
+
+        /// <inheritdoc />
+        public string[] GetGenericDictionaryParameterNames()
+        {
+            var dictionarySymbol = _symbolLookup.GetSymbol(OpenGenericDictionaryTypeInfo);
```

Comment:
Seems wasteful to do this every time. Doesn't this not change in the lifetime of the application?